Index: testhal/STM32/STM32F4xx/IRQ_STORM/iar/ch.ewp
===================================================================
--- testhal/STM32/STM32F4xx/IRQ_STORM/iar/ch.ewp	(revision 7106)
+++ testhal/STM32/STM32F4xx/IRQ_STORM/iar/ch.ewp	(working copy)
@@ -29,8 +29,8 @@
         </option>
         <option>
           <name>Variant</name>
-          <version>19</version>
-          <state>38</state>
+          <version>20</version>
+          <state>39</state>
         </option>
         <option>
           <name>GEndianMode</name>
@@ -61,7 +61,7 @@
         <option>
           <name>FPU</name>
           <version>2</version>
-          <state>0</state>
+          <state>5</state>
         </option>
         <option>
           <name>OGCoreOrChip</name>
@@ -87,7 +87,7 @@
         </option>
         <option>
           <name>OGLastSavedByProductVersion</name>
-          <state>6.30.3.53229</state>
+          <state>6.50.3.4744</state>
         </option>
         <option>
           <name>GeneralEnableMisra</name>
@@ -99,7 +99,7 @@
         </option>
         <option>
           <name>OGChipSelectEditMenu</name>
-          <state>STM32F4xxx	ST STM32F4xxx</state>
+          <state>STM32F407VG	ST STM32F407VG</state>
         </option>
         <option>
           <name>GenLowLevelInterface</name>
@@ -137,13 +137,13 @@
         </option>
         <option>
           <name>GFPUCoreSlave</name>
-          <version>19</version>
-          <state>38</state>
+          <version>20</version>
+          <state>39</state>
         </option>
         <option>
           <name>GBECoreSlave</name>
-          <version>19</version>
-          <state>38</state>
+          <version>20</version>
+          <state>39</state>
         </option>
         <option>
           <name>OGUseCmsis</name>
@@ -163,6 +163,10 @@
         <wantNonLocal>1</wantNonLocal>
         <debug>1</debug>
         <option>
+          <name>CCOptimizationNoSizeConstraints</name>
+          <state>0</state>
+        </option>
+        <option>
           <name>CCDefines</name>
           <state></state>
         </option>
@@ -294,20 +298,24 @@
         <option>
           <name>CCIncludePath2</name>
           <state>$PROJ_DIR$\..\</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\kernel\include</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\ports\common\ARMCMx</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\ports\common\ARMCMx\CMSIS\include</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx\STM32f4xx</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\include</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\DMAv2</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\GPIOv2</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\SPIv1</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\USARTv1</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32f4xx</state>
-          <state>$PROJ_DIR$\..\..\..\..\boards\ST_STM32f4_DISCOVERY</state>
-          <state>$PROJ_DIR$\..\..\..\..\test</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\rt\include</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\rt\ports\ARMCMx</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\ext\CMSIS\ST</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\ext\CMSIS\include</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\rt\ports\ARMCMx\compilers\IAR</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\common\ports\ARMCMx\devices\STM32F4xx</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\include</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\osal\rt</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\platforms\STM32</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\common\ARMCMx</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\LLD\TIMv1</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\DMAv2</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\LLD\GPIOv2</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\LLD\SPIv1</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\LLD\USARTv1</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\STM32F4xx</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\boards\ST_STM32F4_DISCOVERY</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\test</state>
         </option>
         <option>
           <name>CCStdIncCheck</name>
@@ -553,8 +561,9 @@
         <option>
           <name>AUserIncludes</name>
           <state>$PROJ_DIR$\..</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx\STM32f4xx</state>
-          <state>$PROJ_DIR$\..\..\..\..\boards\ST_STM32f4_DISCOVERY</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\common\ports\ARMCMx\devices\STM32F4xx</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\common\ports\ARMCMx\devices\STM32F4xx</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\rt\ports\ARMCMx</state>
         </option>
         <option>
           <name>AExtraOptionsCheckV2</name>
@@ -621,7 +630,7 @@
       <name>ILINK</name>
       <archiveVersion>0</archiveVersion>
       <data>
-        <version>14</version>
+        <version>15</version>
         <wantNonLocal>1</wantNonLocal>
         <debug>1</debug>
         <option>
@@ -790,10 +799,6 @@
           <state>1</state>
         </option>
         <option>
-          <name>CrcAlgo</name>
-          <state>1</state>
-        </option>
-        <option>
           <name>CrcPoly</name>
           <state>0x11021</state>
         </option>
@@ -895,6 +900,16 @@
           <name>IlinkStackCallGraphFile</name>
           <state></state>
         </option>
+        <option>
+          <name>CrcAlgorithm</name>
+          <version>0</version>
+          <state>1</state>
+        </option>
+        <option>
+          <name>CrcUnitSize</name>
+          <version>0</version>
+          <state>0</state>
+        </option>
       </data>
     </settings>
     <settings>
@@ -951,8 +966,8 @@
         </option>
         <option>
           <name>Variant</name>
-          <version>19</version>
-          <state>38</state>
+          <version>20</version>
+          <state>39</state>
         </option>
         <option>
           <name>GEndianMode</name>
@@ -1059,13 +1074,13 @@
         </option>
         <option>
           <name>GFPUCoreSlave</name>
-          <version>19</version>
-          <state>38</state>
+          <version>20</version>
+          <state>39</state>
         </option>
         <option>
           <name>GBECoreSlave</name>
-          <version>19</version>
-          <state>38</state>
+          <version>20</version>
+          <state>39</state>
         </option>
         <option>
           <name>OGUseCmsis</name>
@@ -1085,6 +1100,10 @@
         <wantNonLocal>1</wantNonLocal>
         <debug>0</debug>
         <option>
+          <name>CCOptimizationNoSizeConstraints</name>
+          <state>0</state>
+        </option>
+        <option>
           <name>CCDefines</name>
           <state>NDEBUG</state>
         </option>
@@ -1216,20 +1235,20 @@
         <option>
           <name>CCIncludePath2</name>
           <state>$PROJ_DIR$\..\</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\kernel\include</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\ports\common\ARMCMx</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\ports\common\ARMCMx\CMSIS\include</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx\STM32f4xx</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\include</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\DMAv2</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\GPIOv2</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\SPIv1</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\USARTv1</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32f4xx</state>
-          <state>$PROJ_DIR$\..\..\..\..\boards\ST_STM32f4_DISCOVERY</state>
-          <state>$PROJ_DIR$\..\..\..\..\test</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\rt\include</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\ports\common\ARMCMx</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\ports\common\ARMCMx\CMSIS\include</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\ports\IAR\ARMCMx</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\ports\IAR\ARMCMx\STM32f4xx</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\include</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\platforms\STM32</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\platforms\STM32\DMAv2</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\platforms\STM32\GPIOv2</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\platforms\STM32\SPIv1</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\platforms\STM32\USARTv1</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\hal\platforms\STM32f4xx</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\boards\ST_STM32f4_DISCOVERY</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\test</state>
         </option>
         <option>
           <name>CCStdIncCheck</name>
@@ -1475,8 +1494,8 @@
         <option>
           <name>AUserIncludes</name>
           <state>$PROJ_DIR$\..</state>
-          <state>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx\STM32f4xx</state>
-          <state>$PROJ_DIR$\..\..\..\..\boards\ST_STM32f4_DISCOVERY</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\os\ports\IAR\ARMCMx\STM32f4xx</state>
+          <state>$PROJ_DIR$\..\..\..\..\..\boards\ST_STM32f4_DISCOVERY</state>
         </option>
         <option>
           <name>AExtraOptionsCheckV2</name>
@@ -1543,7 +1562,7 @@
       <name>ILINK</name>
       <archiveVersion>0</archiveVersion>
       <data>
-        <version>14</version>
+        <version>15</version>
         <wantNonLocal>1</wantNonLocal>
         <debug>0</debug>
         <option>
@@ -1712,10 +1731,6 @@
           <state>1</state>
         </option>
         <option>
-          <name>CrcAlgo</name>
-          <state>1</state>
-        </option>
-        <option>
           <name>CrcPoly</name>
           <state>0x11021</state>
         </option>
@@ -1817,6 +1832,16 @@
           <name>IlinkStackCallGraphFile</name>
           <state></state>
         </option>
+        <option>
+          <name>CrcAlgorithm</name>
+          <version>0</version>
+          <state>1</state>
+        </option>
+        <option>
+          <name>CrcUnitSize</name>
+          <version>0</version>
+          <state>0</state>
+        </option>
       </data>
     </settings>
     <settings>
@@ -1849,10 +1874,10 @@
   <group>
     <name>board</name>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\boards\ST_STM32f4_DISCOVERY\board.c</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\os\hal\boards\ST_STM32f4_DISCOVERY\board.c</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\boards\ST_STM32f4_DISCOVERY\board.h</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\os\hal\boards\ST_STM32F4_DISCOVERY\board.h</name>
     </file>
   </group>
   <group>
@@ -1862,127 +1887,127 @@
       <group>
         <name>include</name>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\adc.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\adc.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\can.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\can.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\ext.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\ext.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\gpt.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\gpt.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\hal.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\hal.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\i2c.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\i2c.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\icu.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\icu.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\mac.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\mac.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\mii.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\mii.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\mmc_spi.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\mmc_spi.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\pal.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\pal.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\pwm.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\pwm.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\rtc.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\rtc.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\sdc.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\sdc.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\serial.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\serial.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\serial_usb.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\serial_usb.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\spi.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\spi.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\tm.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\tm.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\uart.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\uart.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\usb.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\usb.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\include\usb_cdc.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\include\usb_cdc.h</name>
         </file>
       </group>
       <group>
         <name>src</name>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\adc.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\adc.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\can.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\can.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\ext.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\ext.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\gpt.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\gpt.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\hal.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\hal.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\i2c.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\i2c.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\icu.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\icu.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\mac.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\mac.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\mmc_spi.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\mmc_spi.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\pal.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\pal.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\pwm.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\pwm.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\rtc.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\rtc.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\sdc.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\sdc.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\serial.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\serial.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\serial_usb.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\serial_usb.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\spi.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\spi.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\tm.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\st.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\uart.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\uart.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\hal\src\usb.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\hal\src\usb.c</name>
         </file>
       </group>
     </group>
@@ -1991,285 +2016,288 @@
       <group>
         <name>include</name>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\ch.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\ch.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chcond.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chcond.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chdebug.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chdebug.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chdynamic.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chdynamic.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chevents.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chevents.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chheap.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chheap.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chinline.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chinline.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chioch.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chioch.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chlists.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chlists.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chmboxes.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chmboxes.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chmemcore.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chmemcore.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chmempools.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chmempools.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chmsg.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chmsg.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chmtx.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chmtx.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chqueues.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chqueues.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chregistry.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chregistry.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chschd.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chschd.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chsem.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chsem.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chstreams.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chstreams.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chsys.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chsys.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chthreads.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chthreads.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\include\chvt.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\include\chvt.h</name>
         </file>
       </group>
       <group>
         <name>src</name>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chcond.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chcond.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chdebug.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chdebug.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chdynamic.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chdynamic.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chevents.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chevents.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chheap.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chheap.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chlists.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chmboxes.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chmboxes.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chmemcore.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chmemcore.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chmempools.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chmempools.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chmsg.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chmsg.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chmtx.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chmtx.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chqueues.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chqueues.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chregistry.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chregistry.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chschd.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chschd.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chsem.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chsem.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chsys.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chsys.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chtm.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chthreads.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chthreads.c</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\kernel\src\chvt.c</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\rt\src\chvt.c</name>
         </file>
       </group>
     </group>
     <group>
       <name>platform</name>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\gpt_lld.c</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\LLD\TIMv1\gpt_lld.c</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\gpt_lld.h</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\LLD\TIMv1\st_lld.c</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32f4xx\hal_lld.c</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\LLD\TIMv1\gpt_lld.h</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32f4xx\hal_lld.h</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\STM32F4xx\hal_lld.c</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\GPIOv2\pal_lld.c</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\STM32F4xx\hal_lld.h</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\GPIOv2\pal_lld.h</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\LLD\GPIOv2\pal_lld.c</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\USARTv1\serial_lld.c</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\LLD\GPIOv2\pal_lld.h</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\USARTv1\serial_lld.h</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\LLD\USARTv1\serial_lld.c</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\SPIv1\spi_lld.c</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\USARTv1\serial_lld.h</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32\SPIv1\spi_lld.h</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\LLD\SPIv1\spi_lld.c</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32f4xx\stm32_dma.c</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\LLD\SPIv1\spi_lld.h</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32f4xx\stm32_dma.h</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\STM32F4xx\stm32_dma.c</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\hal\platforms\STM32f4xx\stm32f4xx.h</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32\STM32F4xx\stm32_dma.h</name>
       </file>
+      <file>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\STM32f4xx\stm32f4xx.h</name>
+      </file>
     </group>
     <group>
       <name>port</name>
       <group>
         <name>STM32F4xx</name>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx\STM32f4xx\cmparams.h</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\ports\IAR\ARMCMx\STM32f4xx\cmparams.h</name>
         </file>
         <file>
-          <name>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx\STM32f4xx\vectors.s</name>
+          <name>$PROJ_DIR$\..\..\..\..\..\os\common\ports\ARMCMx\compilers\IAR\vectors.s</name>
         </file>
       </group>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx\chcore.c</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\rt\ports\ARMCMx\chcore.c</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx\chcore.h</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\rt\ports\ARMCMx\chcore.h</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx\chcore_v7m.c</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\rt\ports\ARMCMx\chcore_v7m.c</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx\chcore_v7m.h</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\rt\ports\ARMCMx\chcore_v7m.h</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx\chcoreasm_v7m.s</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\rt\ports\ARMCMx\compilers\IAR\chcoreasm_v7m.s</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx\chtypes.h</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\ports\IAR\ARMCMx\chtypes.h</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\ports\IAR\ARMCMx\cstartup.s</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\common\ports\ARMCMx\compilers\IAR\cstartup.s</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\ports\common\ARMCMx\nvic.c</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\common\ARMCMx\nvic.c</name>
       </file>
       <file>
-        <name>$PROJ_DIR$\..\..\..\..\os\ports\common\ARMCMx\nvic.h</name>
+        <name>$PROJ_DIR$\..\..\..\..\..\os\hal\ports\common\ARMCMx\nvic.h</name>
       </file>
     </group>
   </group>
   <group>
     <name>test</name>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\test.c</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\test.c</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\test.h</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\test.h</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testbmk.c</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testbmk.c</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testbmk.h</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testbmk.h</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testdyn.c</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testdyn.c</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testdyn.h</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testdyn.h</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testevt.c</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testevt.c</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testevt.h</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testevt.h</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testheap.c</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testheap.c</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testheap.h</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testheap.h</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testmbox.c</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testmbox.c</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testmbox.h</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testmbox.h</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testmsg.c</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testmsg.c</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testmsg.h</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testmsg.h</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testmtx.c</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testmtx.c</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testmtx.h</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testmtx.h</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testpools.c</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testpools.c</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testpools.h</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testpools.h</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testqueues.c</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testqueues.c</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testqueues.h</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testqueues.h</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testsem.c</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testsem.c</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testsem.h</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testsem.h</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testthd.c</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testthd.c</name>
     </file>
     <file>
-      <name>$PROJ_DIR$\..\..\..\..\test\testthd.h</name>
+      <name>$PROJ_DIR$\..\..\..\..\..\test\rt\testthd.h</name>
     </file>
   </group>
   <file>
